scripts/plot_performance_breakdown.py

```diff
@@ -35,31 +40,66 @@
             for statement in metrics['statements']:
                 sum_sql_translation_duration += statement['sql_translation_duration']
                 sum_optimization_duration += statement['optimization_duration']
+                if statement['optimizer_rule_durations']:
+                    for rule_name, rule_duration in statement['optimizer_rule_durations'].items():
+                        if rule_name not in sum_optimizer_rule_durations:
+                            sum_optimizer_rule_durations[rule_name] = rule_duration
+                        else:
+                            sum_optimizer_rule_durations[rule_name] += rule_duration
                 sum_lqp_translation_duration += statement['lqp_translation_duration']
                 sum_plan_execution_duration += statement['plan_execution_duration']
 
     benchmark.append(sum_parse_duration / len(benchmark_json['successful_runs']))
     benchmark.append(sum_sql_translation_duration / len(benchmark_json['successful_runs']))
-    benchmark.append(sum_optimization_duration / len(benchmark_json['successful_runs']))
+    if options.plot_rules:
+        for rule_duration in sum_optimizer_rule_durations.values():
+            benchmark.append(rule_duration / len(benchmark_json['successful_runs']))
+        benchmark.append(0)
+    else:
+        benchmark.append(sum_optimization_duration / len(benchmark_json['successful_runs']))
     benchmark.append(sum_lqp_translation_duration / len(benchmark_json['successful_runs']))
     benchmark.append(sum_plan_execution_duration / len(benchmark_json['successful_runs']))
 
     benchmarks.append(benchmark)
 
-df = pd.DataFrame(benchmarks, columns=['Benchmark', 'Parser', 'SQLTranslator', 'Optimizer', 'LQPTranslator', 'Execution'])
+column_names = ['Benchmark', 'Parser', 'SQLTranslator']
+if options.plot_rules:
+    column_names.extend(sum_optimizer_rule_durations.keys())
+    column_names.append("Other Rules")
+else:
+    column_names.append('Optimizer')
+column_names.extend(['LQPTranslator', 'Execution'])
+df = pd.DataFrame(benchmarks, columns=column_names)
 
+total_time = df.iloc[:,1:].apply(lambda x: x.sum(), axis=1)
 # Normalize data from nanoseconds to percentage of total cost
 df.iloc[:,1:] = df.iloc[:,1:].apply(lambda x: x / x.sum(), axis=1)
 print(df)
 
-ax = df.plot.bar(x='Benchmark', stacked=True)
+if options.plot_rules:
+    # aggregate all rule durations below the threshold
+    threshold = 0.05
+    for index, benchmark in df[sum_optimizer_rule_durations.keys()].iterrows():
+        for rule_name, rule_duration in benchmark.iteritems():
+            if rule_duration < threshold:
+                df.loc[index, "Other Rules"] += rule_duration
```

Comment:
Right now, the order of rules seems random. Can we change that and move "other" to the bottom?

Reply:
indeed. would you prefer alphabetic ordering? ordering them in the actual execution order might be difficult since some rules are applied more than once.

Reply:
In my opinion, the actual execution order does not matter. So anything that's not random is fine.